File: solutions/cpp/0221.cpp
class Solution {
 public:
  int maximalSquare(vector<vector<char>>& matrix) {
    if (matrix.empty()) return 0;

    int maxLength = 0;
    vector<vector<int>> dp(matrix.size(), vector<int>(matrix[0].size()));

    for (int i = 0; i < matrix.size(); ++i)
      for (int j = 0; j < matrix[0].size(); ++j) {
        if (i == 0 || j == 0 || matrix[i][j] == '0')
          dp[i][j] = matrix[i][j] == '1' ? 1 : 0;
        else
          dp[i][j] = min({dp[i - 1][j - 1], dp[i - 1][j], dp[i][j - 1]}) + 1;
        maxLength = max(maxLength, dp[i][j]);
      }

    return maxLength * maxLength;
  }
};

File: solutions/java/0221.java
class Solution {
  public int maximalSquare(char[][] matrix) {
    if (matrix.length == 0)
      return 0;

    int[] dp = new int[matrix[0].length];
    int maxLength = 0;
    int prev = 0;

    for (int i = 0; i < matrix.length; ++i)
      for (int j = 0; j < matrix[0].length; ++j) {
        int cache = dp[j];
        if (i == 0 || j == 0 || matrix[i][j] == '0')
          dp[j] = matrix[i][j] == '1' ? 1 : 0;
        else
          dp[j] = Math.min(dp[j], Math.min(dp[j - 1], prev)) + 1;
        maxLength = Math.max(maxLength, dp[j]);
        prev = cache;
      }

    return maxLength * maxLength;
  }
}